### PR TITLE
disallow_untyped_defs for mypy

### PR DESCRIPTION
**Pull request by ethteck**
Added `disallow_untyped_defs = True` to our mypy rules.

I'd appreciate it if a couple people go over this. Particularly any time I added `Any` or left a `# TODO` I'm less sure about.

### CODE REVIEW

File: backend/coreapp/serializers.py [hunk 1]
```diff
@@ -55,18 +57,18 @@ def serialize_profile(request: Request, profile: Profile, small=False):
...
 if TYPE_CHECKING:
-    ProfileFieldBaseClass = serializers.RelatedField[Profile, str, str]
+    ProfileFieldBaseClass = serializers.RelatedField[Profile, str, dict[str, Any]]
 else:
     ProfileFieldBaseClass = serializers.RelatedField
 
 
 class ProfileField(ProfileFieldBaseClass):
-    def to_representation(self, profile: Profile):
+    def to_representation(self, profile: Profile) -> dict[str, Any]:
```
Comment (nanaian) [inline backend/coreapp/serializers.py:66]: Should this be `Dict`?

File: backend/coreapp/serializers.py [hunk 2]
```diff
@@ -146,23 +150,38 @@ class Meta:
             "platform",
         ]
 
-    def get_project(self, scratch: Scratch):
+    # TODO: Unused
+    def get_project(self, scratch: Scratch) -> Optional[str]:
```
Comment (ethteck) [inline backend/coreapp/serializers.py:154]: @nanaian do you know why these are here? I don't think we need them on the serializer, and they seem unused

File: backend/coreapp/serializers.py [hunk 3]
```diff
@@ -146,23 +150,38 @@ class Meta:
...
+            return reverse(
+                "project-detail",
+                args=[scratch.project_function.project.slug],
+                request=self.context["request"],  # type: ignore
+            )
+        return None
 
-    def get_project_function(self, scratch: Scratch):
+    # TODO: Unused
+    def get_project_function(self, scratch: Scratch) -> Optional[str]:
```
Comment (ethteck) [inline backend/coreapp/serializers.py:167]: this one too

File: backend/coreapp/views/scratch.py [hunk 4]
```diff
@@ -406,12 +406,15 @@ def claim(self, request, pk):
...
     @action(detail=True, methods=["POST"])
-    def fork(self, request, pk):
+    def fork(self, request: Request, pk: str) -> Response:
         parent: Scratch = self.get_object()
 
-        request_data = (
-            request.data.dict() if isinstance(request.data, QueryDict) else request.data
-        )
+        # TODO Needed for test_fork_scratch test?
+        if isinstance(request.data, QueryDict):  # type: ignore
```
Comment (ethteck) [inline backend/coreapp/views/scratch.py:413]: this whole thing seems wrong too

**Review by nanaian — APPROVED** [review]

File: backend/coreapp/serializers.py (hunk 2 above)
Comment (nanaian) [reply backend/coreapp/serializers.py:154]: these are not unused
https://www.django-rest-framework.org/api-guide/fields/#serializermethodfield

**Review by nanaian — CHANGES_REQUESTED** [review]
Remove those `// TODO: Unused` comments; they're not unused

File: backend/coreapp/views/scratch.py (hunk 4 above)
Comment (nanaian) [reply backend/coreapp/views/scratch.py:413]: IIRC when testing, `request.data` was a QueryDict, not a real dict, even though the stubs said it was supposed to be a dict, and the `**request_data` on L419 didn't work with the QueryDict

File: backend/coreapp/serializers.py (hunk 2 above)
Comment (ethteck) [reply backend/coreapp/serializers.py:154]: okay, thanks. They used to not always return a value (the function would fall through to nothingness) - is the `return None` I added at the end of them (and the return type `str`) correct?

Comment (nanaian) [reply backend/coreapp/serializers.py:154]: yep LGTM

File: backend/coreapp/views/scratch.py (hunk 4 above)
Comment (ethteck) [reply backend/coreapp/views/scratch.py:413]: yeah I saw the same thing. I'm just confused why that's the case..

Comment (nanaian) [reply backend/coreapp/views/scratch.py:413]: [it appears that DRF varies the type based on what type of data is POSTed, which is gross](https://stackoverflow.com/a/70230790)

**Review by nanaian — APPROVED** [review]

File: backend/coreapp/tests.py [hunk 5]
```diff
@@ -33,7 +33,7 @@
 from .models.scratch import CompilerConfig, Scratch
 
 
-def requiresCompiler(*compilers: Compiler):
+def requiresCompiler(*compilers: Compiler) -> Callable[..., Any]:
```
Comment (simonlindholm) [inline backend/coreapp/tests.py:36]: not that it matters whatsoever in a test, but you could do `Callable[[F], F]` where `F = TypeVar('F', bound=Callable[..., Any])` for more accurate typing. (The `bound=` part not too important.)

**Review by simonlindholm — APPROVED** [review]
LGTM. The remaining `Any`s look annoying to get rid of, but now they're at least explicit!

**Pull request merged**